### PR TITLE
[DTensor] implement dist_cat as a sharding prop rule

### PR DESCRIPTION
**Pull request by XilunWu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #92677

### CODE REVIEW

File: test/distributed/_tensor/test_dtensor_ops.py [hunk 1]
```diff
@@ -150,7 +150,6 @@ def wrapped(fn):
     xfail("diff"),
     xfail("dist"),
     xfail("dot"),
-    xfail("dstack"),
```
Comment (wanchaol) [inline test/distributed/_tensor/test_dtensor_ops.py:153]: The cat `xfail` still there, should fix bugs and remove that xfail too.

File: test/distributed/_tensor/test_tensor_ops.py [hunk 2]
```diff
@@ -354,6 +354,24 @@ def test_index(self):
...
+        torch.manual_seed(self.rank)
+        tensor_1 = torch.rand(3, 5, 6)
+        tensor_2 = torch.rand(3, 5, 6)
+        tensor_3 = torch.rand(3, 5, 6)
+        sharding = [Shard(0)]
+        dt_1 = DTensor.from_local(tensor_1, device_mesh, sharding)
+        dt_2 = DTensor.from_local(tensor_2, device_mesh, sharding)
+        dt_3 = DTensor.from_local(tensor_3, device_mesh, sharding)
+        new_dt = torch.cat([dt_1, dt_2, dt_3])
+        cat_dt = DTensor.from_local(
```
Comment (wanchaol) [inline test/distributed/_tensor/test_tensor_ops.py:369]: this test is actually wrong, we should either fix it or remove it. We should rely more on the op db tests to test our op implementation (i.e. remove the xfail of cat)

File: torch/distributed/_tensor/ops/tensor_ops.py [hunk 3]
```diff
@@ -472,3 +472,93 @@ def place(vp: Placement, ip: Placement) -> Placement:
...
+    if (len(op_schema.args_schema) > 1):
+        dim = cast(int, op_schema.args_schema[1])
+    # normalize arguments
+    if dim < 0:
+        dim += tensor_list_specs[0].ndim
+
+    # check concat dim
+    needs_reshard_on_cat_dim = False
+    for spec in tensor_list_specs:
+        if dim < len(spec.placements) and spec.placements[dim].is_shard():
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/tensor_ops.py:490]: This logic is wrong, the `needs_reshard_on_cat_dim` should only be set when the DTensor's shard dim (tensor dim) is the same as the cat_dim, but you are checking the mesh dim instead of the tensor dim here! you can use spec.dim_map[dim] >=0 to check instead

File: torch/distributed/_tensor/ops/tensor_ops.py [hunk 4]
```diff
@@ -472,3 +472,93 @@ def place(vp: Placement, ip: Placement) -> Placement:
...
+            needs_reshard_on_cat_dim = True
+            spec.placements = unshard_tensor_dim(spec.placements, dim=dim)
+    if needs_reshard_on_cat_dim:
+        args_schema = (tensor_list_specs,) + op_schema.args_schema[1:]
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/tensor_ops.py:494]: what is this? the suggested_schema is the same as the input schema? shouldn't it be a schema suggestion to reshard the input to a legit schema?

File: torch/distributed/_tensor/ops/tensor_ops.py [hunk 5]
```diff
@@ -472,3 +472,93 @@ def place(vp: Placement, ip: Placement) -> Placement:
...
+        suggested_schema = OpSchema(
+            func_schema=op_schema.func_schema,
+            args_schema=args_schema,
+            kwargs_schema=op_schema.kwargs_schema,
+        )
+        return OutputSharding(
+            None,
+            schema_suggestions=[suggested_schema],
+            failed_reason="All tensors in concat must have no sharding on cat dim, need to reshard!",
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/tensor_ops.py:503]: It's not failing so you don't need to specify the failed reason

File: torch/distributed/_tensor/ops/tensor_ops.py [hunk 6]
```diff
@@ -472,3 +472,93 @@ def place(vp: Placement, ip: Placement) -> Placement:
...
+        )
+    alphabet = "abcdefghijklmnopqrstuvwxyz"
+    einop_equation = ""
+    for spec in tensor_list_specs:
+        einop_equation += alphabet[:spec.ndim]
+        einop_equation += ','
+    einop_equation = einop_equation[:-1] + "->" + alphabet[:tensor_list_specs[0].ndim]
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/tensor_ops.py:510]: this special "-1" is too hacky. please not use it. Instead, you can create a list and hold all the dim_chars, then do a join in the end, see https://github.com/pytorch/pytorch/blob/400a00b94955b2b188c04ff261b3ff0fe9e21e74/torch/distributed/_tensor/ops/common_rules.py#L268 as an example.

File: torch/distributed/_tensor/ops/tensor_ops.py [hunk 7]
```diff
@@ -472,3 +472,93 @@ def place(vp: Placement, ip: Placement) -> Placement:
...
+
+    if output_sharding.output_spec is None:
+        if output_sharding.schema_suggestions is not None:
+            return _update_schema_suggestion_for_cat(
+                output_sharding,
+                op_schema,
+                dim,
+            )
+        else:
+            return OutputSharding(None)
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/tensor_ops.py:529]: shouldn't you return the `output_sharding` directly instead of creating a new one?

File: torch/distributed/_tensor/ops/tensor_ops.py [hunk 8]
```diff
@@ -472,3 +472,93 @@ def place(vp: Placement, ip: Placement) -> Placement:
...
+    assert isinstance(output_sharding.output_spec, DTensorSpec)
+    output_sharding.output_spec.shape = torch.Size(
+        tuple(output_sharding.output_spec.shape[:dim])
+        + (new_size,)
+        + tuple(output_sharding.output_spec.shape[dim + 1 :])
+    )
+    return output_sharding
+
+
+def _update_schema_suggestion_for_cat(
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/tensor_ops.py:543]: can you tell me what exactly this function is doing? it looks like a lot of duplicate logic with the rule itself and I am not quite sure what this function is used for.

File: torch/distributed/_tensor/ops/tensor_ops.py [hunk 9]
```diff
@@ -472,3 +472,93 @@ def place(vp: Placement, ip: Placement) -> Placement:
...
+    # normalize arguments
+    if dim < 0:
+        dim += tensor_list_specs[0].ndim
+
+    # check concat dim
+    needs_reshard_on_cat_dim = False
+    for spec in tensor_list_specs:
+        if dim < len(spec.placements) and spec.placements[dim].is_shard():
+            needs_reshard_on_cat_dim = True
+            spec.placements = unshard_tensor_dim(spec.placements, dim=dim)
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/tensor_ops.py:492]: please don't modify the original schema and its specs inplace, create a new schema with the expected placements.

**Review by wanchaol — CHANGES_REQUESTED** [review]
See comments inlined, we should make sure the `xfail` of cat to be removed so that it passes all the possible cases.

File: torch/distributed/_tensor/ops/tensor_ops.py (hunk 8 above)
Comment (XilunWu) [reply torch/distributed/_tensor/ops/tensor_ops.py:594]: `einop_rule` expects the `op_schema` argument to have its `args_schema` in form `[DTensorSpec, DTensorSpec, ...]` but when it's passed into `cat_rule` the schema is actually `[List[DTensorSpec]]`. That's why I convert the `args_schema` at the beginning of `cat_rule` (https://github.com/pytorch/pytorch/pull/92677/files#diff-ebc7be1151cf411ce7edf46c4ca1cabb74cd953a2bdf47e04b4cc733c31f6085R492) before feeding it into `einop_rule`. Thus, we need to convert it back if a `schema_suggestion` is present here.

File: torch/distributed/_tensor/ops/tensor_ops.py [hunk 10]
```diff
@@ -472,3 +484,128 @@ def place(vp: Placement, ip: Placement) -> Placement:
...
+        + (new_size,)
+        + tuple(output_sharding.output_spec.shape[dim + 1 :])
+    )
+    return output_sharding
+
+
+def _update_schema_suggestion_for_cat(
+    output_sharding: OutputSharding,
+    op_schema: OpSchema,
+    dim: int,
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/tensor_ops.py:597]: where is the `dim` being used?

File: torch/distributed/_tensor/ops/tensor_ops.py [hunk 11]
```diff
@@ -472,3 +484,128 @@ def place(vp: Placement, ip: Placement) -> Placement:
...
+
+    l = len(tensor_list_specs)
+    free_dim = alphabet[l:l + ndim - 1]
+    for i, spec in enumerate(tensor_list_specs):
+        if spec.ndim == ndim:
+            # rewrite concat dim
+            dim_word = free_dim[:dim] + alphabet[i] + free_dim[dim:]
+            einop_notation_list.append(dim_word)
+        else:
+            einop_notation_list.append(alphabet[i])
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/tensor_ops.py:541]: is this the empty tensor annotation where it have a single char?

**Review by wanchaol — APPROVED** [review]
lgtm, thanks for working on it! left a couple of suggestions and some question.

File: torch/distributed/_tensor/ops/tensor_ops.py (hunk 10 above)
Comment (XilunWu) [reply torch/distributed/_tensor/ops/tensor_ops.py:597]: good catch!

File: torch/distributed/_tensor/ops/tensor_ops.py (hunk 11 above)
Comment (XilunWu) [reply torch/distributed/_tensor/ops/tensor_ops.py:541]: Not entirely for empty tensor but empty tensor whose `ndim` is smaller than other tensors. This is for case like concatenating `Tensor([], shape=torch.Size([0]))` with `Tensor([[1, 2], [3, 4]], shape=torch.Size([2, 2]))`.
